torch_xla/csrc/runtime/computation_client.h

```diff
@@ -64,31 +70,67 @@ class ComputationClient {
 
   using DataPtr = std::shared_ptr<Data>;
 
-  class Computation {
+  // There are 4 different Computation class being used here
+  // 1. torch::lazy::Computation represent a general computation from LTC
+  // perspective.
+  // 2. runtime::ComputationClient::Computation represent a computation from the
+  // ComputationClient perspective. It wraps a xla::XlaComputation and a vector
+  // of device.
+  // 3. xla::XlaComputation represent a xla computation, it is generated by the
+  // xla compiler.
+  // 4. xla::PjRtComputationClient::PjRtComputation which inherits from
+  // runtime::ComputationClient::Computation and contains a handle to represent
+  // the compiled program.
+  class Computation : public torch::lazy::Computation {
    public:
-    Computation(xla::XlaComputation computation,
-                xla::ProgramShape program_shape,
-                std::vector<std::string> devices)
-        : computation_(std::move(computation)),
-          program_shape_(std::move(program_shape)),
-          devices_(std::move(devices)) {}
-
-    Computation(xla::XlaComputation computation)
-        : computation_(std::move(computation)) {
+    // Our Computation is being used for 3 different purpose.
+    // 1. To represent a xla computation build by xla_op_builder, in which case
+    //    we would need the name and hash. Computation would be a wrapper around
+    //    a runtime::ComputationClient::Computation.
+    //    runtime::ComputationClient::Computation::devices_ would be empty.
+    // 2. To represent a computation built by syncTensor and needs to be
+    //    compiled.
+    //    ...
+    Computation(std::string name, xla::XlaComputation computation,
+                std::vector<std::string> devices = {})
+        : name_(name),
+          computation_(std::move(computation)),
+          devices_(std::move(devices)) {
       program_shape_ = ConsumeValue(computation_.GetProgramShape());
+      hash_ =
+          torch::lazy::MHash(name, computation_.proto().SerializeAsString());
     }
 
+    Computation(std::string name, xla::XlaComputation computation,
+                torch::lazy::BackendDevice device)
+        : Computation(name, std::move(computation), {device.toString()}) {}
+
+    // ...
+    // 3. To represent a computation that is already compiled. In this case
+    //    name_ and hash_ are not required. Computation will be a wrapper around
+    //    an executable, PjRtComputationClient::PjRtComputation in our case. It
+    //    is not ideal to use same class for 3 different purposes but this is
+    //    the path took by upstream ltc.
     Computation(xla::XlaComputation computation,
                 std::vector<std::string> devices)
-        : computation_(std::move(computation)), devices_(std::move(devices)) {
-      program_shape_ = ConsumeValue(computation_.GetProgramShape());
-    }
```

Comment:
any reason `program_shape_` is only being inited on one of the constructors?

Reply:
All of the other constructors delegate to the one that sets `program_shape_` and `hash_`.

https://secure.wikimedia.org/wikipedia/en/wiki/C++11#Object_construction_improvement